### PR TITLE
Emit warning when TestEngine ID starts with "junit-"

### LINKED ISSUE #1557: Check TestEngine ID for "junit-" prefix on registration
## Overview

The `junit-` prefix is owned by the JUnit Team. It's akin to using `org.junit` as a base package. No other entity in the world should do that.

Only two engines should use the prefix at the moment:

* `junit-jupiter`
* `junit-vintage`

Source of the discussion: https://github.com/junit-team/junit5/commit/5a44da8c6145611057923a8ae46df540c566fc04#r30200049

## Deliverables

- [x] Document in the User Guide that the JUnit Team _owns_ the `junit-` prefix for `TestEngine` IDs.
- [x] Emit a warning when a `TestEngine` ID starts with `junit-`, except those published by the JUnit Team from the JUnit 5 repository for public consumption.
- [x] Create follow-up ticket for ignoring all such _foreign_ engines in a future release of the Platform: #1572

### PR DESCRIPTION
**Pull request by sormuras**
## Overview

The `junit-` prefix is owned by the JUnit Team. It's akin to using **"org.junit"** as a base package. No other entity in the world should do that. Only two engines should use the prefix at the moment:

 - `junit-jupiter`
 - `junit-vintage`

This commit enforces these rules.

Closes #1557

---

I hereby agree to the terms of the [JUnit Contributor License Agreement](https://github.com/junit-team/junit5/blob/002a0052926ddee57cf90580fa49bc37e5a72427/CONTRIBUTING.md#junit-contributor-license-agreement).

---

### Definition of Done

- [x] There are no TODOs left in the code
- [x] Method [preconditions](http://junit.org/junit5/docs/snapshot/api/org/junit/platform/commons/util/Preconditions.html) are checked and documented in the method's Javadoc
- [x] [Coding conventions](https://github.com/junit-team/junit5/blob/master/CONTRIBUTING.md#coding-conventions) (e.g. for logging) have been followed
- [x] Change is covered by [automated tests](https://github.com/junit-team/junit5/blob/master/CONTRIBUTING.md#tests) including corner cases, errors, and exception handling
- [x] Public API has [Javadoc](https://github.com/junit-team/junit5/blob/master/CONTRIBUTING.md#javadoc) and [`@API` annotations](http://junit.org/junit5/docs/snapshot/api/org/junit/platform/commons/meta/API.html)
- [ ] Change is documented in the [User Guide](http://junit.org/junit5/docs/snapshot/user-guide/) and [Release Notes](http://junit.org/junit5/docs/snapshot/user-guide/#release-notes)
- [x] All [continuous integration builds](https://github.com/junit-team/junit5#continuous-integration-builds) pass

### CODE REVIEW

File: junit-platform-launcher/src/main/java/org/junit/platform/launcher/core/DefaultLauncher.java [hunk 1]
```diff
@@ -58,12 +58,19 @@
...
+		this.testEngines = validateEngineIds(testEngines);
 	}
 
-	private static Iterable<TestEngine> validateUniqueIds(Iterable<TestEngine> testEngines) {
+	private static Iterable<TestEngine> validateEngineIds(Iterable<TestEngine> testEngines) {
 		Set<String> ids = new HashSet<>();
 		for (TestEngine testEngine : testEngines) {
+			// check usage of reserved id prefix
+			if (!validateReservedIds(testEngine)) {
+				logger.warn(() -> String.format("Thou shalt not use an engine ID starting with 'junit-': '%s'",
```
Comment (sbrannen) [inline junit-platform-launcher/src/main/java/org/junit/platform/launcher/core/DefaultLauncher.java]: The style here is too biblical.

Let's opt for something more like: _Third-party TestEngine implementations are forbidden to use the reserved `junit-` prefix for their ID._

File: junit-platform-launcher/src/main/java/org/junit/platform/launcher/core/DefaultLauncher.java [hunk 2]
```diff
@@ -72,6 +79,32 @@
...
+		return false;
+	}
+
+	private static void validateWellKnownTypeName(TestEngine testEngine, String expectedTypeName) {
+		String actualTypeName = testEngine.getClass().getTypeName();
+		if (actualTypeName.equals(expectedTypeName)) {
+			return;
+		}
+		throw new JUnitException(
+			String.format("Impostor detected! Expected %s but got: %s", expectedTypeName, actualTypeName));
```
Comment (sbrannen) [inline junit-platform-launcher/src/main/java/org/junit/platform/launcher/core/DefaultLauncher.java]: Hos about something more like: _Third-party TestEngine [FQCN] is forbidden to use the reserved 'junit-vintage' TestEngine ID._

**Comment by sormuras** [comment]
Requested changes applies. Documentation is pending.
